# Практическое занятие 3
# Компьютерный практикум по алгебре на Python
## Решение систем линейных алгебраических уравнений (СЛАУ)

https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linsolve

In [85]:
import sympy
from sympy import linsolve, Matrix, S, Symbol, symbols, Eq,\
linear_eq_to_matrix, simplify, ones
from IPython.display import Latex

### Задание 1.
Решить СЛАУ, представив их тремя различными способами (расширенная матрица, список уравнений, матричный вид)
$$
a)\ \left\{\begin{matrix} - 2 x_{1} + 3 x_{2} - 9 x_{3} = 51 \\ 4 x_{1} - 6 x_{2} + x_{3} = 51 \\ 5 x_{1} + 6 x_{2} - 8 x_{3} = 93\end{matrix}\right.\qquad
b)\ \left\{\begin{matrix} 2 x_{1} + 6 x_{2} - x_{3} + 2 x_{4} = -9 \\ - 5 x_{1} - x_{2} + 5 x_{3} + 2 x_{4} = -98 \\ - x_{1} + 2 x_{2} - 2 x_{3} + 7 x_{4} = -41\end{matrix}\right.\qquad
c)\ \left\{\begin{matrix} - 6 x_{1} - x_{2} + x_{3} = -59 \\ 8 x_{1} + x_{2} + x_{3} = 59 \\ 2 x_{1} - 6 x_{2} + 2 x_{3} = 25 \\ 2 x_{1} - 7 x_{2} + 9 x_{3} = -35\end{matrix}\right.\qquad
$$

In [ ]:
Ab = Matrix([[-2, 3, -9, 51], [4, -6, 1, 51], [5, 6, -8, 93]])
display(linsolve(Ab))

x1, x2, x3, x4 = symbols('x1:5')
eqns = [Eq(2*x1 + 6*x2 - x3 + 2*x4, -9),
        Eq(-5*x1 - x2 + 5*x3 + 2*x4, -98),
        Eq(-x1 + 2*x2 -2*x3 + 7*x4, -41)]
B, b = linear_eq_to_matrix(eqns, [x1, x2, x3, x4])
display(linsolve((B, b)))

C = Matrix([[-6, -1, 1], [8, 1, 1], [2, -6, 2], [2, -7, 9]])
b = Matrix([-59, 59, 25, -35])
display(linsolve((C, b)))

{(9, -4, -9)}

{(207*tau0/95 + 2097/95, -15*tau0/19 - 166/19, 154*tau0/95 + 69/95, tau0)}

EmptySet

**Ответ:**
$$
a)\ A = \left[\begin{matrix}-2 & 3 & -9\\4 & -6 & 1\\5 & 6 & -8\end{matrix}\right],\ b = \left[\begin{matrix}51\\51\\93\end{matrix}\right],\ X = \left\{\left( 9, \  -4, \  -9\right)\right\}\\
b)\ A = \left[\begin{matrix}2 & 6 & -1 & 2\\-5 & -1 & 5 & 2\\-1 & 2 & -2 & 7\end{matrix}\right],\ b = \left[\begin{matrix}-9\\-98\\-41\end{matrix}\right],\ X = \left\{\left( \frac{207 \tau_{0}}{95} + \frac{2097}{95}, \  - \frac{15 \tau_{0}}{19} - \frac{166}{19}, \  \frac{154 \tau_{0}}{95} + \frac{69}{95}, \  \tau_{0}\right)\right\}\\
c)\ A = \left[\begin{matrix}-6 & -1 & 1\\8 & 1 & 1\\2 & -6 & 2\\2 & -7 & 9\end{matrix}\right],\ b = \left[\begin{matrix}-59\\59\\25\\-35\end{matrix}\right],\ X = \emptyset\\
$$

### Задание 2.

Решить СЛАУ с параметром $a$, представив их списком уравнений, а затем приведя к матричному виду с помощью linear_eq_to_matrix
$$
a)\ \left\{
\begin{matrix}
x+3y-2az=8\\
2x-4y+5z=-1\\
3x+4y+5z=10
\end{matrix}
\right., \quad
b)\ \left\{
\begin{matrix}
x+5(a-1)y-3z=8\\
x-(5a + 4)y+4z=-1\\
3x+99y+5z=6
\end{matrix}
\right., \quad
c) \left\{
\begin{matrix}
-x+5ay-3z=1\\
4ax-ay+z= -5\\
(120a^2 - 6a)x + (20a^2 - a) y+z= - 140 a - 5
\end{matrix}
\right.
$$
Получить с помощью linsolve решение с произвольным значением параметра (т.е. подставить матрицу как есть, с символами $a$).
Решить СЛАУ а) при $a=-\frac{25}{40}$ и $a=-0.625$, b) при $a$, обращающем в 0 определитель, записанном в символьном и десятичном формате, c) при $a=0,\ S(1) / 20,\ 0.05$.

Проанализировать результаты. Для каждого случая (кроме случая несовместной СЛАУ) провести проверку подстановкой.

При подстановке дробей (символьный формат) пользоваться конструкцией вида $S(num)/den$, например,
$S(2)/3$

In [ ]:
a, x, y, z = symbols('a x y z')
eqns = [Eq(x + 3*y - 2*a*z, 8), Eq(2*x - 4*y + 5*z, -1), Eq(3*x + 4*y + 5*z, 10)]
A, b = linear_eq_to_matrix(eqns, [x, y, z])
X = linsolve((A, b))
display(X)
X_matr = Matrix(*X)
display(simplify(A * X_matr - b).T)
# подстановка в а) -25/40
X = linsolve((A.subs(a, -S(25)/40), b))
display(X) # решений нет
# подстановка в а) -0.625
X = linsolve((A.subs(a, (-0.625)), b))
display(X)
X_matr = Matrix(*X)
display(simplify(A.subs(a, -0.625) * X_matr - b).T)

{((72*a + 155)/(40*a + 25), (46*a + 15)/(40*a + 25), -11/(8*a + 5))}

Matrix([[0, 0, 0]])

EmptySet

{(-1.98158383604302e+16, 2.47697979505377e+15, 9.90791918021509e+15)}

Matrix([[-4.0, -7.0, -10]])

In [ ]:
eqns = [Eq(x + 5*(a - 1)*y - 3*z, 8), Eq(x - (5*a + 4)*y + 4*z, -1), Eq(3*x + 99*y + 5*z, 6)]
B, b = linear_eq_to_matrix(eqns, [x, y, z])
B_det = B.det()
display(B_det)
solution = solve(B_det, a)
display(*solution)

-35*a - 784

-112/5

In [ ]:
eqns = [Eq(-x + 5*a*y - 3*z, 1), Eq(4*a*x - a*y + z, -5), Eq((120*a**2 - 6*a)*x + (20*a**2 - a)*y + z, -140*a - 5)]
C, b = linear_eq_to_matrix(eqns, [x, y, z])
X = linsolve((C, b))
display(X)
X_matr = Matrix(*X)
display(simplify(C * X_matr - b).T)
# подстановка в с) 0
X = linsolve((C.subs(a, 0), b.subs(a, 0)))
display(X)
X_matr = Matrix(*X)
display((C.subs(a, 0)*X_matr - b.subs(a, 0)).T)
# подстановка в с) S(1)/20
X = linsolve((C.subs(a, 𝑆(1)/20), b.subs(a, 𝑆(1)/20)))
display(X)
X_matr = Matrix(*X)
display((C.subs(a, 𝑆(1)/20)*X_matr - b.subs(a, 𝑆(1)/20)).T)
# подстановка в с) 0.05
X = linsolve((C.subs(a, 0.05), b.subs(a, 0.05)))
display(X)
X_matr = Matrix(*X)
display((C.subs(a, 0.05) * X_matr - b.subs(a, 0.05)).T)

{(-2*tau0/(20*a - 1) - 24/(20*a - 1), tau0*(12*a - 1)/(20*a**2 - a) + (4*a - 5)/(20*a**2 - a), tau0)}

Matrix([[0, 0, 0]])

{(14, tau0, -5)}

Matrix([[0, 0, 0]])

{(tau0/4 + 35, tau0, -12)}

Matrix([[0, 0, 0]])

{(64.0, 256.0, -12.0)}

Matrix([[35.0, -7.0, 0]])

### Задание 3
В  матрице
$$
 \left[\begin{matrix}-2 & 3 & -9 & 5\\
 4 & -6 & 1 & -2\\
 5 & 6 & -3 & 0
 \end{matrix}\right]
$$

1) выделить матрицу из правых двух столбцов

2) выделить нижнюю строку

3) транспонировать матрицу, полученную в 1)

4) получить расширенную матрицу некоторой СЛАУ из матрицы 3) (как матрицы левой части) и столбца из единиц

5) создать на основе лямбда-функции матрицу $3\times 4$ из расположенных в шахматном порядке чисел 2 и 3, в левом верхнем углу 2.

In [ ]:
A = Matrix([[-2, 3, -9, 5], [4, -6, 1, -2], [5, 6, -3, 0]])
display(A.extract([0, 1, 2], [0, 1]))
display(A[2, 0:4])
display((A.extract([0, 1, 2], [0, 1])).T)
b = Matrix([1, 1])
display(((A.extract([0, 1, 2], [0, 1])).T).row_join(b))
P = Matrix(3, 4, lambda i, j: 2 if (i+j) % 2 == 0 else 3)
display(P)

Matrix([
[-2,  3],
[ 4, -6],
[ 5,  6]])

Matrix([[5, 6, -3, 0]])

Matrix([
[-2,  4, 5],
[ 3, -6, 6]])

Matrix([
[-2,  4, 5, 1],
[ 3, -6, 6, 1]])

Matrix([
[2, 3, 2, 3],
[3, 2, 3, 2],
[2, 3, 2, 3]])

### Задание 4* четверг
В расширенной матрице СЛАУ найти значения параметра $b$, при котором она имеет единственное решение, не имеет решений,  бесконечное число решений.

Для каждого случая составить СЛАУ, найти решение с помощью linsolve и вывести на экран значение параметра и найденное решение СЛАУ.
$$
\left\{
\begin{matrix}
x+2y+bz=3\\
-x-4y+5z=-1\\
bx+2y+z=3
\end{matrix}
\right.
$$

In [ ]:
from sympy import solve
x, y, z, b = symbols('x y z b')
Ab = Matrix([[1, 2, b, 3], [-1, -4, 5, -1], [b, 2, 1, 3]])
A = Matrix([[1, 2, b], [-1, -4, 5], [b, 2, 1]])
A_det = A.det()
sol1 = solve(A_det, b)
display(sol1)
# при b = -3 - нет решений
display(linsolve(Ab.subs(b, -3)))
# при b = 1 - бесконечно много решений
display(linsolve(Ab.subs(b, 1)))
# при b != -3, 1 - одно решение
display(linsolve(Ab))

[-3, 1]

EmptySet

{(5 - 7*tau0, 3*tau0 - 1, tau0)}

{(5/(2*b + 6), (b + 13)/(4*b + 12), 5/(2*b + 6))}

### Задание 5*
Описать функцию с обязательными аргументами матрицей левой части СЛАУ и матрицей столбцом правой части, возвращающую список, первый элемент которого - решение СЛАУ (может быть пустым множеством!), второй - текст о том, сколько решений у СЛАУ ("единственное решение", "бесконечное число решений", "нет решений").  

In [97]:
def f(matr, b):
  Ab = matr.row_join(b)
  solution = (linsolve(Ab))
  matr_det = A.det()
  if (matr_det != 0):
    return [solution, "единственное решение"]
  elif len(solution) == 0:
    return [solution, "нет решений"]
  return [solution, "бесконечное число решений"]

**Сложность задачи состоит в определении ситуации бесконечного числа решений.**

### Индивидуальное задание.
СЛАУ с параметром задана матрицей левой части и столбцом правой части. Представить СЛАУ расширенной матрицей и списком уравнений (автоматически, изученными средствами sympy), вывести на экран.

Решить СЛАУ c параметром тремя способами, провести проверку подстановкой (один раз).

Затем отдельно рассмотреть значение параметра, при котором решение СЛАУ нельзя получить по общей формуле, полученной ранее.
Найти решение СЛАУ при этом значении параметра первым или третьим способом, используя подстановку subs.

In [96]:
b, x1, x2, x3, x4 = symbols('b x1:5')
A = Matrix([[2, -2, 1, -7], [4, 0, 6, 1], [b, -6, -6, -4], [-2, -2, -5, -8]])
B = Matrix([-54, 28, -80, -82])
Ab = A.row_join(B)
display(Ab)
X = Matrix([x1, x2, x3, x4])
AX = A*X
display(AX - B)
# решение СЛАУ 3-мя способами
Ab = Matrix([[2, -2, 1, -7, -54], [4, 0, 6, 1, 28], [b, -6, -6, -4, -80], [-2, -2, -5, -8, -82]])
display(linsolve(Ab))

eqns = [Eq(2*x1 -2*x2 + x3 - 7*x4, -54),
        Eq(4*x1 + 0*x2 + 6*x3 + x4, 28),
        Eq(b*x1 - 6*x2 - 6*x3 - 4*x4, -80),
        Eq(-2*x1 -2*x2 -5*x3 -8*x4, -82)]
A, B = linear_eq_to_matrix(eqns, [x1, x2, x3, x4])
display(linsolve((A, B)))

A = Matrix([[2, -2, 1, -7], [4, 0, 6, 1], [b, -6, -6, -4], [-2, -2, -5, -8]])
B = Matrix([-54, 28, -80, -82])
display(linsolve((A, B)))
sol1 = linsolve(Ab)
X = Matrix(*sol1)
display(simplify(A*X - B).T)
# осбобый случай при b = 0
A = Matrix([[2, -2, 1, -7], [4, 0, 6, 1], [b, -6, -6, -4], [-2, -2, -5, -8]])
A = A.subs(b, 0)
B = Matrix([-54, 28, -80, -82])
display(linsolve((A, B)))

Matrix([
[ 2, -2,  1, -7, -54],
[ 4,  0,  6,  1,  28],
[ b, -6, -6, -4, -80],
[-2, -2, -5, -8, -82]])

Matrix([
[   2*x1 - 2*x2 + x3 - 7*x4 + 54],
[          4*x1 + 6*x3 + x4 - 28],
[ b*x1 - 6*x2 - 6*x3 - 4*x4 + 80],
[-2*x1 - 2*x2 - 5*x3 - 8*x4 + 82]])

{(-37*tau0/(2*b) + 124/b, tau0*(-43*b - 148)/(12*b) + (88*b + 248)/(3*b), tau0*(74 - b)/(6*b) + (14*b - 248)/(3*b), tau0)}

{(-37*tau0/(2*b) + 124/b, tau0*(-43*b - 148)/(12*b) + (88*b + 248)/(3*b), tau0*(74 - b)/(6*b) + (14*b - 248)/(3*b), tau0)}

{(-37*tau0/(2*b) + 124/b, tau0*(-43*b - 148)/(12*b) + (88*b + 248)/(3*b), tau0*(74 - b)/(6*b) + (14*b - 248)/(3*b), tau0)}

Matrix([[0, 0, 0, 0]])

{(197/37 - 3*tau0/2, 328/37 - tau0, tau0, 248/37)}